### PR TITLE
Adds initial support for hot reload for Fuchsia to flutter_tool.

### PR DESCRIPTION
**Pull request by zanderso**

### CODE REVIEW

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 1]
```diff
@@ -87,8 +87,75 @@ class HotRunner extends ResidentRunner {
...
+    }
+
+    if (stayResident)
+      return waitForAppToFinish();
+    await cleanupAtFinish();
+  }
+
   @override
-  Future<int> run({
+  Future<int> _run({
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/run_hot.dart]: I believe this is a bad merge and it should be run

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 2]
```diff
@@ -366,6 +375,7 @@ abstract class ResidentRunner {
 
   void printHelpDetails() {
     if (supportsServiceProtocol) {
+      printStatus('To detach from the application without ending it, press "d".');
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/resident_runner.dart]: I'm not sure we want to show this to every Flutter developer. Maybe make this line optional depending on Fuchsia?

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 3]
```diff
@@ -273,6 +279,9 @@ abstract class ResidentRunner {
       // F10, exit
       await stop();
       return true;
+    } else if (lower == 'd') {
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Why is this needed?

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  final String _fuchsiaRoot;
+  final Random _rng = new Random(new DateTime.now().millisecondsSinceEpoch);
+
+  FuchsiaCommandRunner(this._fuchsiaRoot);
+
+  Future<List<String>> run(String command) async {
+    int tag = _rng.nextInt(999999);
+    String netruncmd = "$_fuchsiaRoot/out/build-magenta/tools/netruncmd";
+    String netcp = "$_fuchsiaRoot/out/build-magenta/tools/netcp";
+    String remoteFile = "/tmp/netruncmd.$tag";
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: you need to use final for all of these variables

Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: remoteFile -> remoteStdout

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    String localFile = "${Directory.systemTemp.path}/netruncmd.$tag";
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: localFile -> localStdout

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+// TODO(zra): When Fuchsia has ssh, this should be changed to use that instead.
+class FuchsiaCommandRunner {
+  final String _fuchsiaRoot;
+  final Random _rng = new Random(new DateTime.now().millisecondsSinceEpoch);
+
+  FuchsiaCommandRunner(this._fuchsiaRoot);
+
+  Future<List<String>> run(String command) async {
+    int tag = _rng.nextInt(999999);
+    String netruncmd = "$_fuchsiaRoot/out/build-magenta/tools/netruncmd";
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: here and below:

kNetRunCommand = 'out/build-magenta/tools/netruncmd';

fs.path.join(_fuchsiaRoot, kNetRunCommand).

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+  bool _fileExists(String path) {
+    File f = new File(path);
+    return f.existsSync();
+  }
+}
+
+
+// TODO(zra): When Fuchsia has ssh, this should be changed to use that instead.
+class FuchsiaCommandRunner {
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: maybe rename this class to be "FuchsiaDeviceCommandRunner"

**Review by johnmccutchan — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 3 above)
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Apps on Fuchsia don't shut down cleanly after a 'q'. This is useful maybe only until I track that down and/or while bringing up hot reload for Fuchsia. I'll leave this here and remove the help message.

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 1 above)
Comment (zanderso) [inline packages/flutter_tools/lib/src/run_hot.dart]: Yah. Done.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 2 above)
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Removed.

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart (hunk 7 above)
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: Done.

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart (hunk 6 above)
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: Done.

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart (hunk 4 above)
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: Done.

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart (hunk 5 above)
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: Done.

**Comment by zanderso** [comment]
PTAL

**Review by johnmccutchan — APPROVED** [review]

**Comment by johnmccutchan** [comment]
LGTM
